kernel/src/hil/adc.rs

```diff
@@ -1,38 +1,80 @@
+/// Interfaces for analog to digital converter peripherals.
 use returncode::ReturnCode;
 
-/// Trait for handling callbacks from ADC module.
-pub trait Client {
-    /// Called when a sample is ready.
-    fn sample_done(&self, sample: u16);
-}
 
-/// Simple interface for reading a single ADC sample on any channel.
-pub trait AdcSingle {
+// *** Interfaces for low-speed, single-sample ADCs ***
+
+/// Simple interface for reading an ADC sample on any channel.
+pub trait Adc {
+    /// The chip-dependent type of an ADC channel.
+    type Channel;
+
     /// Initialize must be called before taking a sample.
-    /// Returns true on success.
     fn initialize(&self) -> ReturnCode;
 
     /// Request a single ADC sample on a particular channel.
-    /// Returns true on success.
-    fn sample(&self, channel: u8) -> ReturnCode;
-    fn cancel_sample(&self) -> ReturnCode;
+    /// Used for individual samples that have no timing requirements.
+    fn sample(&self, channel: &Self::Channel) -> ReturnCode;
+
+    /// Request repeated ADC samples on a particular channel.
+    /// Callbacks will occur at the given frequency with low jitter and can be set to any frequency
+    /// supported by the chip implementation. However callbacks may be limited based on how quickly
+    /// the system can service individual samples, leading to missed samples at high frequencies.
+    fn sample_continuous(&self, channel: &Self::Channel, frequency: u32) -> ReturnCode;
+
+    /// Stop a sampling operation.
+    /// Can be used to stop simple or high-speed samples. No further callback will occur in simple
+    /// mode. The `sampling_complete` callback will occur in high-speed mode.
+    fn stop_sampling(&self) -> ReturnCode;
```

Comment:
In the current design in this PR, when you call `stop_sampling` you will either receive no callbacks (single sampling mode) or always receive a `sampling_complete` callback (high speed mode) within the same call stack. Which is not quite re-entrant since you'll likely call it from a `samples_ready` callback, but certainly a second callback before the first has finished.

In practice, all I the `sampling_complete` callback does is store the returned buffers in takecells and set active to false. (https://github.com/helena-project/tock/blob/adc-continuous/capsules/src/adc.rs#L666) But I can certainly understand how those interactions could be extremely confusing for someone who hadn't just written both sides of the interface.

What would your suggestion be for a different way to handle the interface?

One idea that comes to mind is that instead of ever having a callback return the buffers, I could have a function call to the ADC that retrieves them. Something like:
`retrieve_buffers() -> (Option<&mut [u16]>, Option<&mut[u16]>)`
which would return the buffers from the lower-level ADC if it isn't running.

Alternatively, I could have `stop_sampling` return the buffers, probably a tuple of:
`(ReturnCode, Options<&mut [u16]>, Option<&mut [u16]>)`
but then I would need to update `sample_highspeed` to have the same return type (in case an error occurs and we need to get the buffers back to the user). Might be a better interface though. In this model, `stop_sampling` would still always succeed, but no additional callbacks would ever occur.

Reading back through this post, I think this last option (returning a tuple including the buffers) sounds like the best interface. But I would be happy to hear your thoughts.